src/objective/rank_obj.cu

```diff
@@ -0,0 +1,637 @@
+      int pos_idx = -1;  // Bigger label
+      int neg_idx = -1;  // Smaller label
+      // Are we picking a sample to the left/right of the current group?
+      if (sample < nleft) {
+        // Go left
+        pos_idx = sample + group_begin;
+        neg_idx = item_idx;
+      } else {
+        pos_idx = item_idx;
+        int items_in_group = total_group_items - nleft - nright;
+        neg_idx = sample + items_in_group + group_begin;
+      }
+
+      // Compute and assign the gradients now
+      const float eps = 1e-16f;
+      bst_float p = common::Sigmoid(dpreds[original_pos[pos_idx]] - dpreds[original_pos[neg_idx]]);
+      bst_float g = p - 1.0f;
+      bst_float h = thrust::max(p * (1.0f - p), eps);
+
+      // Rescale each gradient and hessian so that the group has a weighted constant
+      float scale = 1.0f / (niter / total_items);
+      if (fix_list_weight != 0.0f) {
+        scale *= fix_list_weight / total_group_items;
+      }
+
+      float weight = num_weights ? dweights[group_idx - 1] : 1.0f;
+      weight *= weight_normalization_factor;
+      weight *= scale;
+      // Accumulate gradient and hessian in both positive and negative indices
+      const GradientPair in_pos_gpair(g * weight, 2.0f * weight * h);
+      dh::AtomicAddGpair(&out_gpair[original_pos[pos_idx]], in_pos_gpair);
+
+      const GradientPair in_neg_gpair(-g * weight, 2.0f * weight * h);
+      dh::AtomicAddGpair(&out_gpair[original_pos[neg_idx]], in_neg_gpair);
+    });
+  }
+};
+#endif
+
+// objective for lambda rank
+template <typename LambdaWeightComputerT>
+class LambdaRankObj : public ObjFunction {
+ public:
+  void Configure(const std::vector<std::pair<std::string, std::string> >& args) override {
+    param_.InitAllowUnknown(args);
+  }
+
+  void GetGradient(const HostDeviceVector<bst_float>& preds,
+                   const MetaInfo& info,
+                   int iter,
+                   HostDeviceVector<GradientPair>* out_gpair) override {
+    CHECK_EQ(preds.Size(), info.labels_.Size()) << "label size predict size not match";
+
+    // quick consistency when group is not available
+    std::vector<unsigned> tgptr(2, 0); tgptr[1] = static_cast<unsigned>(info.labels_.Size());
+    const std::vector<unsigned> &gptr = info.group_ptr_.size() == 0 ? tgptr : info.group_ptr_;
+    CHECK(gptr.size() != 0 && gptr.back() == info.labels_.Size())
+        << "group structure not consistent with #rows";
+
+    const auto ngroup = static_cast<bst_omp_uint>(gptr.size() - 1);
+    bst_float sum_weights = 0;
+    for (bst_omp_uint k = 0; k < ngroup; ++k) {
+      sum_weights += info.GetWeight(k);
+    }
+    bst_float weight_normalization_factor = ngroup / sum_weights;
+
+#if defined(__CUDACC__)
+    // For now, we only support pairwise ranking computation on GPU.
+    // Check if we have a GPU assignment; else, revert back to CPU
+    auto device = tparam_->gpu_id;
+    if (device >= 0 && LambdaWeightComputerT::SupportOnGPU()) {
```

Comment:
There is too much code in this if statement, maybe extract each branch into a method e.g. `GetGradientCPU`/`GetGradientGPU`

Reply:
done